satpy/readers/ascat_l2_soilmoisture_bufr.py

```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# Copyright (c) 2019 Satpy developers
+#
+# This file is part of satpy.
+#
+# satpy is free software: you can redistribute it and/or modify it under the
+# terms of the GNU General Public License as published by the Free Software
+# Foundation, either version 3 of the License, or (at your option) any later
+# version.
+#
+# satpy is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# satpy.  If not, see <http://www.gnu.org/licenses/>.
+r"""ASCAT Soil moisture product reader for BUFR messages.
+Based on the IASI L2 SO2 BUFR reader
+
+"""
+
+import logging
+from datetime import datetime
+import numpy as np
+import xarray as xr
+import dask.array as da
+
+try:
+    import eccodes as ec
+except ImportError as e:
+    raise ImportError(
+        """Missing eccodes-python and/or eccodes C-library installation. Use conda to install eccodes.
+           Error: """, e)
+
+from satpy.readers.file_handlers import BaseFileHandler
+from satpy import CHUNK_SIZE
+
+logger = logging.getLogger('ASCATSOILMOISTUREBUFR')
+
+
+class ASCATSOILMOISTUREBUFR(BaseFileHandler):
+    """File handler for the ASCAT Soil Moisture BUFR product."""
+
+    def __init__(self, filename, filename_info, filetype_info, **kwargs):
+        """Initialise the file handler for the ASCAT Soil Moisture BUFR data."""
+        super(ASCATSOILMOISTUREBUFR, self).__init__(filename, filename_info, filetype_info)
+
+        start_time, end_time = self.get_start_end_date()
+
+        self.metadata = {}
+        self.metadata['start_time'] = start_time
+        self.metadata['end_time'] = end_time
+
+    @property
+    def start_time(self):
+        """Return the start time of data acqusition."""
+        return self.metadata['start_time']
+
+    @property
+    def end_time(self):
+        """Return the end time of data acquisition."""
+        return self.metadata['end_time']
+
+    @property
+    def platform_name(self):
+        """Return spacecraft name."""
+        return self.filename_info['platform']
+
+    def get_start_end_date(self):
+        """Gets the first and last date from the bufr file."""
+        with open(self.filename, 'rb') as fh:
+            date_min = None
+            date_max = None
+            while True:
+                # get handle for message
+                bufr = ec.codes_bufr_new_from_file(fh)
+                if bufr is None:
+                    break
+                ec.codes_set(bufr, 'unpack', 1)
```

Comment:
It will read the whole file but the file is usually small.  The only metadata is the start / end date.  The bufr data (channels) are optional and I don't think it fits with the metadata so I would read this separately even if it means scanning the file twice.